src/ProjectOrigin.VerifiableEventStore/Services/BlockFinalizer/BlockFinalizerBackgroundService.cs

```diff
@@ -0,0 +1,54 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using ProjectOrigin.VerifiableEventStore.Models;
+using ProjectOrigin.VerifiableEventStore.Services.BlockPublisher;
+using ProjectOrigin.VerifiableEventStore.Services.Repository;
+using ProjectOrigin.VerifiableEventStore.Services.TransactionStatusCache;
+
+namespace ProjectOrigin.VerifiableEventStore.Services.BlockFinalizer;
+
+public class BlockFinalizerBackgroundService : BackgroundService
+{
+    private readonly TimeSpan _blockInterval;
+    private readonly ILogger<BlockFinalizerBackgroundService> _logger;
+    private readonly IServiceProvider _serviceProvider;
+
+    public BlockFinalizerBackgroundService(ILogger<BlockFinalizerBackgroundService> logger, IServiceProvider serviceProvider, IOptions<BlockFinalizationOptions> options)
+    {
+        _logger = logger;
+        _serviceProvider = serviceProvider;
+        _blockInterval = options.Value.Interval;
+    }
+
+    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        var processor = new BlockFinalizerJob(
+            _serviceProvider.GetRequiredService<ILogger<BlockFinalizerJob>>(),
+            _serviceProvider.GetRequiredService<IBlockPublisher>(),
+            _serviceProvider.GetRequiredService<ITransactionRepository>(),
+            _serviceProvider.GetRequiredService<ITransactionStatusService>()
+            );
+
+        using var timer = new PeriodicTimer(_blockInterval);
+
+        while (!stoppingToken.IsCancellationRequested &&
+               await timer.WaitForNextTickAsync(stoppingToken))
```

Comment:
Is there a chance that when the tick occurs on the timer, the stoppingToken has been cancelled, but the code will continue?

Reply:
I don't think so, if the stoppingToken is canceled, then it will break the while loop, and the stoppingtoken is also send to the waitForNext, so if cancelled, then it should also stop it.

Reply:
I think `!stoppingToken.IsCancellationRequested` is executed first, then `await timer.WaitForNextTickAsync(stoppingToken)` will be executed. If `await timer.WaitForNextTickAsync(stoppingToken)` stops due to the stoppingToken being cancelled, then `!stoppingToken.IsCancellationRequested` is not executed again. 
But I am not sure about the exact behaviour.